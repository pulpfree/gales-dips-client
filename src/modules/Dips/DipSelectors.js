import React, { Component } from 'react'
import PropTypes from 'prop-types'

import classNames from 'classnames'
import moment from 'moment'

import ArrowBack from '@material-ui/icons/ArrowBack'
import ArrowForward from '@material-ui/icons/ArrowForward'
import Button from '@material-ui/core/Button'
import FormControl from '@material-ui/core/FormControl'
import { withStyles } from '@material-ui/core/styles'

import MomentUtils from 'material-ui-pickers/utils/moment-utils'
import MuiPickersUtilsProvider from 'material-ui-pickers/utils/MuiPickersUtilsProvider'
import DatePicker from 'material-ui-pickers/DatePicker'

import StationSelector from '../Common/StationSelector'
import { STD_DATE_FORMAT as dateFormat } from '../../config/constants'

class DipSelectors extends Component {
  constructor(props) {
    super(props)
    this.handleStationChange = this.handleStationChange.bind(this)

    this.state = {
      nextDisabled: true,
      selectedDate: moment(),
      stationID: '',
    }
  }

  componentDidMount = () => {
    const { params } = this.props.match
    const { date, stationID } = params
    const setDate = moment(date)
    const newDte = {
      year: setDate.format('YYYY'),
      month: setDate.subtract(1, 'months').format('MM'),
      date: setDate.format('DD'),
    }
    if (date) {
      this.setState({ selectedDate: this.state.selectedDate.set(newDte) }, this.setNextDisabled)
    }
    if (stationID) {
      this.setState({ stationID }, this.handleGetDip)
    }
  }

  // This could get more involved if state isn't set in time: https://stackoverflow.com/questions/37401635/react-js-wait-for-setstate-to-finish-before-triggering-a-function
  handleStationChange = (value) => {
    this.setState({ stationID: value }, this.handleGetDip)
  }

  handleDateChange = (date) => {
    this.setState({ selectedDate: date }, this.handleGetDip)
  }

  handleNextPrevDate = (val) => {
    const dte = this.state.selectedDate
    if (val === 'p') {
      dte.subtract(1, 'days')
    } else if (val === 'n') {
      dte.add(1, 'days')
    }
    this.setState({ selectedDate: dte }, this.handleGetDip)
  }

  handleGetDip = () => {
    this.props.actions.errorClear()
    this.setNextDisabled()
    const { history } = this.props
    const { selectedDate, stationID } = this.state

    if (stationID) {
      const dte = selectedDate.format(dateFormat)
      const uri = `/dips/${dte}/${stationID}`
      history.push(uri)
    }
  }

  setNextDisabled = () => {
    const { selectedDate } = this.state
    const today = moment().format(dateFormat)
    this.setState({ nextDisabled: !selectedDate.isBefore(today) })
  }

  render() {
    const { classes } = this.props
    const { nextDisabled, selectedDate, stationID } = this.state

    return (
      <div className={classes.container}>
        <div className={classes.selectRow}>

          <div className={classes.cell}>
            <FormControl className={classes.formControl}>
              <StationSelector
                onStationChange={this.handleStationChange}
                stationID={stationID}
              />
            </FormControl>
          </div>

          <div className={classNames([classes.cell], [classes.calendar])}>
            <FormControl className={classes.formControl}>
              <MuiPickersUtilsProvider utils={MomentUtils}>
                <DatePicker
                  autoOk
                  disableFuture
                  format="MMM D, YYYY"
                  label="Date"
                  onChange={this.handleDateChange}
                  value={selectedDate}
                />
              </MuiPickersUtilsProvider>
            </FormControl>
          </div>

          <div className={classes.cell}>
            <Button
              className={classes.navButton}
              color="secondary"
              onClick={() => this.handleNextPrevDate('p')}
              variant="fab"
            >
              <ArrowBack />
            </Button>
            <Button
              className={classes.navButton}
              color="secondary"
              disabled={nextDisabled}
              onClick={() => this.handleNextPrevDate('n')}
              variant="fab"
            >
              <ArrowForward />
            </Button>
          </div>

        </div>
      </div>
    )
  }
}

DipSelectors.propTypes = {
  actions: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
}

const styles = theme => ({
  cell: {
    flex: 'flex-grow',
    alignSelf: 'flex-end',
  },
  container: {
    marginBottom: theme.spacing.unit,
  },
  formControl: {
    margin: theme.spacing.unit,
    minWidth: 120,
  },
  navButton: {
    margin: theme.spacing.unit,
    height: theme.spacing.unit * 4.5,
    width: theme.spacing.unit * 4.5,
    marginRight: 0,
  },
  selectRow: {
    display: 'inline-flex',
    flexDirection: 'row',
    marginBottom: theme.spacing.unit,
  },
})

export default withStyles(styles)(DipSelectors)
