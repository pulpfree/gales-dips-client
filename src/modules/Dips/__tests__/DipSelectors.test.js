/* eslint no-undef: "off" */

import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import { shallow } from 'enzyme'

import DipSelectors from '../DipSelectors'

describe('DipSelectors', () => {

  let mountedC
  let props
  const reportC = () => {
    if (!mountedC) {
      mountedC = shallow(
        <Router><DipSelectors {...props} /></Router>
      )
    }
    return mountedC
  }

  beforeEach(() => {
    props = {
      history: {},
      location: {},
      match: {},
      // client: {}
    }
    mountedC = undefined
  })

  /*it('always renders a div', () => {
    const divs = reportC().find('div')
    expect(divs.length).toBeGreaterThan(0)
  })*/

  it('matches snapshot', () => {
    expect(reportC()).toMatchSnapshot()
  })

})