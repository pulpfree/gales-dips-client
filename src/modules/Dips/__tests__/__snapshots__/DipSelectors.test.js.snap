// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DipSelectors matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(DipSelectors)
    actions={Object {}}
    history={Object {}}
    location={Object {}}
    match={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "actions": Object {},
      "classes": Object {
        "cell": "DipSelectors-cell-1",
        "container": "DipSelectors-container-2",
        "formControl": "DipSelectors-formControl-3",
        "navButton": "DipSelectors-navButton-4",
        "selectRow": "DipSelectors-selectRow-5",
      },
      "history": Object {},
      "location": Object {},
      "match": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "actions": Object {},
        "classes": Object {
          "cell": "DipSelectors-cell-1",
          "container": "DipSelectors-container-2",
          "formControl": "DipSelectors-formControl-3",
          "navButton": "DipSelectors-navButton-4",
          "selectRow": "DipSelectors-selectRow-5",
        },
        "history": Object {},
        "location": Object {},
        "match": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
